### ISSUE TITLE
[New Package]: KDE Nightly

### ISSUE THREAD

**Issue by Trenly** [vedantmgoyal2009/winget-pkgs-automation#185]
It seems there are a lot of requests to add the nightly versions of the KDE Packages. These currently do not exist in WinGet.
However, it is a point of contention on whether they should be added or not because wingetbot cannot update them, and because the links to them become invalid after only a few days. 

I believe that the automation in this repository can be used to keep them updated nightly, removing the old versions and replacing them with the new ones. There is a fair amount of discussion in https://github.com/microsoft/winget-pkgs/pull/33718

I'm suggesting that we determine if these packages can be automated. If they can be, then we should add them to winget-pkgs and add the automation here at the same time

Packages which have been requested:
- [ ] microsoft/winget-pkgs#33704
- [ ] microsoft/winget-pkgs#33705
- [ ] microsoft/winget-pkgs#33706
- [ ] microsoft/winget-pkgs#33709
- [ ] microsoft/winget-pkgs#33710
- [ ] microsoft/winget-pkgs#33711
- [ ] microsoft/winget-pkgs#33712
- [ ] microsoft/winget-pkgs#33713
- [ ] microsoft/winget-pkgs#33715
- [ ] microsoft/winget-pkgs#33716
- [x] microsoft/winget-pkgs#33719
- [ ] microsoft/winget-pkgs#33721
- [x] microsoft/winget-pkgs#33722
- [ ] microsoft/winget-pkgs#33725
- [x] microsoft/winget-pkgs#33726
- [ ] microsoft/winget-pkgs#33727
- [ ] microsoft/winget-pkgs#33729
- [ ] microsoft/winget-pkgs#33730
- [ ] microsoft/winget-pkgs#33735
- [ ] microsoft/winget-pkgs#33736
- [ ] microsoft/winget-pkgs#33737
- [ ] microsoft/winget-pkgs#34830
- [ ] microsoft/winget-pkgs#34831
- [ ] microsoft/winget-pkgs#34832

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Much of the problemacy that is relevant to automatic replacement of these packages has been caused by KDE's distributor and compiler, but potentially equally as much has been caused by the inadequate automation that is being used by winget for automatic replacement of packages. Although I have not ascertained how this is able to be improved, because many package-managers are able to manage these packages, evaluation of those may be assistive for remediation of this.

**Comment by vedantmgoyal2009** [comment]
I'll be able to add them after 16th December, since I am currently busy with exams.

**Comment by Trenly** [comment]
> Much of the problemacy that is relevant to automatic replacement of these packages has been caused by KDE's distributor and compiler

I can't agree more. The best way would be for KDE to have a vanity link that gets updated on every build. But, thats neither here nor  there

> but potentially equally as much has been caused by the inadequate automation that is being used by winget for automatic replacement of packages. 

Certainly improvements can be made on both ends. I can say that much of MSFT's time has been dedicated to making the core functionality of the package manager better. This includes things like the 1.1.0 manifest updates and package pinning, to help users avoid issues like upgrade loops or mismatched app entries. They have been working on other things like support for .zip files, support for standalone or portable apps, binaries, drivers, and even PowerShell install scripts. When you consider the relatively small size of the MSFT team that is working on this (compared to other MSFT endeavors), it makes a lot more sense for them to focus on functionality than on automation, at least for where they are in the lifecycle of the project. 

> Although I have not ascertained how this is able to be improved, many package-managers are able to manage these packages, so evaluation of those may be assistive for remediation of this.

The main functionality that I know the bots over at MSFT have in terms of updating packages deals with packages that have vanity URLs. This means that there is a URL which always points to the latest version, so that websites never have to be updated, and so the URL is clean looking. The automation MSFT has built so far constantly checks every single `InstallerUrl: ` in the repository to ensure it results in a 2xx or 3xx->...->2xx status code. If it results in a 4xx or a 5xx status code, the automation creates an issue that the installer could not be downloaded, and the package may need to be removed. If it results in a 2xx status code, the automation downloads the installer and checks that the `InstallerHash` matches what is in the manifest for that installer. If it does not match, it attempts to get the version information from the downloaded installer, and submits a PR to update the manifest.

The main "issue" with the automation (and I use the term "issue" loosely here) is that it cannot automatically update packages where the `InstallerUrl` changes. The most common cause of this is the inclusion of a version number in the URL. The most prevalent examples of this are the packages which are shipped through GitHub Releases, which is what Vedant's repository was originally created around. Because the version numbers are semi-arbitrary, it is an extremely difficult task for automation to find them deterministically. However, if you take a look around this repository, you will see that all the packages that we have included here have a version method and/or a custom script for updating them. All of the packages in this repository have _some method_ that we can call to parse out the version numbers, or get the URL to the latest version which isn't covered by MSFT's automation. For the GitHub releases, we call the GitHub API. For some packages like [Badlion.BadlionClient](https://github.com/vedantmgoyal2009/winget-pkgs-automation/blob/main/packages/b/badlion.badlionclient.json) we are downloading a file with their version information. In some cases, we even parse through HTML code if the download page for the package is easy enough to work with. 

-----
TL;DR -
* Microsoft has other priorities and limited time / resources
* There is not really an easy way to standardize updating all packages
* This repository is a way for the community to contribute to automating package updates until MSFT implements additional methods for package updates

**Comment by vedantmgoyal2009** [comment]
So, I've started adding the packages to the automation.

> I'm suggesting that we determine if these packages can be automated. If they can be, then we should add them to winget-pkgs and add the automation here at the same time

Would anyone like to draft PRs so these packages so that can be added to the repo at the same time?

**Comment by vedantmgoyal2009** [comment]
Nightly releases are currently blocked by http://bugs.kde.org/show_bug.cgi?id=444894, or maybe I will try installing the package and getting ARP entries from it, but I don't think it will be a good approach.

**Comment by Trenly** [comment]
@vedantmgoyal2009 It seems the jenkins API you're using for kde packages currently isn't updating fast enough to capture all the latest releases. I think this method might work to capture all of them, although you'll have to figure out how to get the version number -
```
PS >  $BaseUrl = 'https://binary-factory.kde.org/view/Windows%2064-bit/job/kdeconnect-kde_Release_win64/'
PS >  ((Invoke-WebRequest -Uri "$BaseUrl").Links | Where-Object {$_.outerText -like "Last successful build*"} | Select-Object -Property outerText) -match '[0-9]{3,}' | Out-Null; $build = $Matches.Values
PS >  $BaseUrl+$build+"/$(((Invoke-WebRequest -Uri "$BaseUrl$build/").Links | Where-Object {$_.href -like '*.exe'}).href)" 
https://binary-factory.kde.org/view/Windows%2064-bit/job/kdeconnect-kde_Release_win64/821/artifact/kdeconnect-kde-21.12.0-821-windows-msvc2019_64-cl.exe
PS >  
PS >  $BaseUrl = 'https://binary-factory.kde.org/view/Windows%2064-bit/job/Cantor_Release_win64/'
PS >  ((Invoke-WebRequest -Uri "$BaseUrl").Links | Where-Object {$_.outerText -like "Last successful build*"} | Select-Object -Property outerText) -match '[0-9]{3,}' | Out-Null; $build = $Matches.Values
PS >  $BaseUrl+$build+"/$(((Invoke-WebRequest -Uri "$BaseUrl$build/").Links | Where-Object {$_.href -like '*.exe'}).href)"
https://binary-factory.kde.org/view/Windows%2064-bit/job/Cantor_Release_win64/394/artifact/cantor-21.12.0-394-windows-msvc2019_64-cl.exe
PS >
PS >  $BaseUrl = 'https://binary-factory.kde.org/view/Windows%2064-bit/job/KBibTeX_Nightly_win64/'
PS >  ((Invoke-WebRequest -Uri "$BaseUrl").Links | Where-Object {$_.outerText -like "Last successful build*"} | Select-Object -Property outerText) -match '[0-9]{3,}' | Out-Null; $build = $Matches.Values
PS >  $BaseUrl+$build+"/$(((Invoke-WebRequest -Uri "$BaseUrl$build/").Links | Where-Object {$_.href -like '*.exe'}).href)"
https://binary-factory.kde.org/view/Windows%2064-bit/job/KBibTeX_Nightly_win64/787/artifact/kbibtex-master-787-windows-msvc2019_64-cl.exe
```
